modules/ismp/clients/polygon/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(unused_variables)]
+
+extern crate alloc;
+
+use alloc::{boxed::Box, collections::BTreeMap, string::ToString, vec, vec::Vec};
+use codec::{Decode, Encode};
+use ics23::HostFunctionsManager;
+use ismp::{
+	consensus::{
+		ConsensusClient, ConsensusClientId, ConsensusStateId, StateCommitment, StateMachineClient,
+		StateMachineId, VerifiedCommitments,
+	},
+	error::Error,
+	host::{IsmpHost, StateMachine},
+	messaging::StateCommitmentHeight,
+};
+
+use evm_state_machine::EvmStateMachine;
+use pallet_ismp_host_executive::Config as HostExecutiveConfig;
+use prost::Message;
+use scale_info::prelude::time::{SystemTime, UNIX_EPOCH};
+use tendermint_primitives::{CodecConsensusProof, CodecTrustedState, Milestone};
+use tendermint_verifier::verify_header_update;
+
+/// The consensus update/proof for Polygon
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct PolygonConsensusUpdate {
+	/// Serialized Tendermint light client update (signed header, validator set, etc.)
+	pub tendermint_proof: CodecConsensusProof,
+	/// Milestone update
+	pub milestone_update: Option<MilestoneUpdate>,
+}
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct MilestoneUpdate {
+	/// EVM block header for the milestone's end block
+	pub evm_header: geth_primitives::CodecHeader,
+	/// Milestone number
+	pub milestone_number: u64,
+	/// ICS23 proof for the milestone inclusion
+	pub ics23_state_proof: ICS23Proof,
+	// Milestone
+	pub milestone: Milestone,
+	/// Untrusted header app hash for verification
+	pub untrusted_header_app_hash: Vec<u8>,
```

Comment:
Not needed use what is in the untrusted header from the consensus proof. Basically when verifying proofs, the onchain verifier must use only trusted values for root hashes, keys and values.
```suggestion
```